src/Spork/App.purs

```diff
@@ -109,166 +95,97 @@ type PureApp s i = App (Const Void) (Const Void) s i
 
 type BasicApp m s i = App m (Const Void) s i
 
-type Push eff i = i → Eff eff Unit
-
-newtype Interpreter eff m q i = Interpreter (Push eff i → Coproduct m q i → Eff eff Unit)
-
-newtype Step m i = Step (i → m (Loop m i))
-
-data Loop m i = Loop (i → m (Loop m i)) (m (Step m i))
-
-type EventQueue m i = (i → m Unit) → m (Step m i)
-
-runEventQueue
-  ∷ ∀ eff i
-  . EventQueue (Eff (ref ∷ REF | eff)) i
-  → Eff (ref ∷ REF | eff) (i → Eff (ref ∷ REF | eff) Unit)
-runEventQueue proc = do
-  queue   ← newRef Nothing
-  machine ← newRef Nothing
-
+makeApp
+  ∷ ∀ eff m q s i
+  . Functor m
+  ⇒ Functor q
+  ⇒ (Interpreter (AppEffects eff) m q)
+  → App m q s i
+  → DOM.Node
+  → EventQueue (Eff (AppEffects eff)) (Either (Coproduct m q (Eff (AppEffects eff) Unit)) i)
+makeApp (Interpreter interpreter) app el = makeEventQueue \push →
   let
-    push i = do
-      q ← readRef queue
-      case q of
-        Nothing → do
-          writeRef queue (Just [i])
-          start
-        Just is →
-          writeRef queue (Just (Array.snoc is i))
-
-    start = do
-      step ← readRef machine
-      for_ step (loop <<< startLoop)
-
-    startLoop (Step fn) =
-      Loop fn (pure (Step fn))
+    pushInput = push <<< Right
```

Comment:
These have nothing to do with success/failure, right? What is the reasoning for splitting it into Input vs effect/sub this way?